googletakeout.cc

```diff
@@ -0,0 +1,416 @@
+   */
+  if (false && activitySegment.contains(SIMPLE_PATH)) {
+    const QJsonArray points =
+      activitySegment[SIMPLE_PATH][POINTS].toArray();
+    for (const auto&& pointRef : points) {
+      ++ n_points;
+      const QJsonObject point = pointRef.toObject();
+      timestamp = point[TIMESTAMP].toString();
+      waypoint = _waypoint(
+        point[LATE7].toInt(),
+        point[LONE7].toInt(),
+        nullptr,
+        nullptr,
+        &timestamp
+      );
+      track_maybe_add_wpt(route, waypoint);
+    }
+  } else {
+    const QJsonArray points =
+      activitySegment[WAYPOINT_PATH][WAYPOINTS].toArray();
+    for (const auto&& pointRef: points) {
+      ++ n_points;
+      const QJsonObject point = pointRef.toObject();
+      /* as waypointPath does not include timestamps, set every
+       * point along the path to the start of the activity
+       */
+      waypoint = _waypoint(
+        point[LATE7].toInt(),
+        point[LONE7].toInt(),
+        nullptr,
+        nullptr,
+        &timestamp
+      );
+      track_maybe_add_wpt(route, waypoint);
+    }
+  }
+  timestamp = activitySegment[DURATION][END_TIMESTAMP].toString();
+  waypoint = _waypoint(
+    endLoc[LOCATION_LATE7].toInt(),
+    endLoc[LOCATION_LONE7].toInt(),
+    nullptr, nullptr,
+    &timestamp
+  );
+  track_maybe_add_wpt(route, waypoint);
+  return n_points;
+}
+
+GoogleTakeoutInputStream::GoogleTakeoutInputStream() {}
+
+GoogleTakeoutInputStream::GoogleTakeoutInputStream(const QString& source) {
+  this->sources = { source };
+}
+
+QList<QJsonObject> GoogleTakeoutInputStream::readJson(const QString& source) {
+  Debug(1) << "Reading from JSON " << source;
+  auto* ifd = new gpsbabel::File(source);
+  ifd->open(QIODevice::ReadOnly | QIODevice::Text);
+  const QString content = ifd->readAll();
+  QJsonParseError error{};
+  const QJsonDocument doc = QJsonDocument::fromJson(content.toUtf8(), &error);
+  if (error.error != QJsonParseError::NoError) {
+    _fatal(
+      QString("JSON parse error in ") + ifd->fileName() + ": " +
+      error.errorString()
+    );
+  }
+
+  const QJsonObject root = doc.object();
+  const QJsonValue timelineObjectsIn = root.value(TIMELINE_OBJECTS);
+  if (timelineObjectsIn.isNull()) {
+    _fatal(
+      QString(ifd->fileName()) + " is missing required \"" +
+      TIMELINE_OBJECTS + "\" section"
+    );
+  }
+
+  const QJsonArray timelineJson = timelineObjectsIn.toArray();
+  QList<QJsonObject> timeline;
+  for (QJsonValue&& val : timelineJson) {
+    if (val.isObject()) {
+      timeline.append(val.toObject());
+    } else {
+      _fatal(QString(ifd->fileName()) + " has non-object in timelineObjects");
+    }
+  }
+  ifd->close();
+  delete ifd;
+  if (timeline.isEmpty()) {
+    _warning(QString(source) + " does not contain any timelineObjects");
+  }
+  Debug(1) << "Saw " << timeline.size() << " timelineObjects in " << source;
+  return timeline;
+}
+
+QList<QString> GoogleTakeoutInputStream::readDir(const QString& source) {
```

Comment:
warning: method 'readDir' can be made static [readability-convert-member-functions-to-static]